File: 강의 자료/02-알고리즘/05~06-이분 탐색/문제별 코드/13702-이상한 술집/solution.cpp
#include <iostream>
#include <algorithm>
#include <string>
typedef long long int ll;
using namespace std;

#define NM 10005

int N, K;
int A[NM];

void input() {
	cin >> N >> K;
	for (int i = 1; i <= N; i++) cin >> A[i];
}

bool determination(ll amount) {
	if (amount == 0) {
		return false;
	}
	ll sum = 0;
	for (int i = 1; i <= N; i++) {
		sum += A[i] / amount;
	}
	return sum >= K;
}

void pro() {
	ll L = 0, R = 1ll << 31, ans = 0;
	while (L <= R) {
		int mid = (L + R) / 2;
		if (determination(mid)) {
			ans = mid;
			L = mid + 1;
		}
		else R = mid - 1;
	}
	cout << ans;
}


int main() {
	ios::sync_with_stdio(false); cin.tie(0); cout.tie(0);
	input();
	pro();
	return 0;
}

File: 강의 자료/02-알고리즘/05~06-이분 탐색/문제별 코드/13702-이상한 술집/solution.py
import sys
si = sys.stdin.readline
n, k = list(map(int, si().split()))
a = [int(si()) for _ in range(n)]

def determination(amount):
    if amount == 0:
        return False
    sum = 0
    for x in a:
        sum += x // amount
    return sum >= k

l, r, ans = 0, 1 << 31, 0
while l <= r:
    mid = (l + r) // 2
    if determination(mid):
        ans = mid
        l = mid + 1
    else:
        r = mid - 1

print(ans)
